syncapi/streams/stream_pdu.go

```diff
@@ -204,59 +234,46 @@ func (p *PDUStreamProvider) addRoomDeltaToResponse(
 	return nil
 }
 
-func (p *PDUStreamProvider) getJoinResponseForCompleteSync(
+func (p *PDUStreamProvider) getResponseForCompleteSync(
 	ctx context.Context,
 	roomID string,
 	r types.Range,
-	stateFilter *gomatrixserverlib.StateFilter,
-	numRecentEventsPerRoom int, device *userapi.Device,
-) (jr *types.JoinResponse, err error) {
-	var stateEvents []*gomatrixserverlib.HeaderedEvent
-	stateEvents, err = p.DB.CurrentState(ctx, roomID, stateFilter)
-	if err != nil {
-		return
-	}
+	filter *gomatrixserverlib.Filter,
+	device *userapi.Device,
+) (
+	recentEvents, stateEvents []*gomatrixserverlib.HeaderedEvent,
+	prevBatch *types.TopologyToken, limited bool, err error,
+) {
 	// TODO: When filters are added, we may need to call this multiple times to get enough events.
 	//       See: https://github.com/matrix-org/synapse/blob/v0.19.3/synapse/handlers/sync.py#L316
 	var recentStreamEvents []types.StreamEvent
-	var limited bool
 	recentStreamEvents, limited, err = p.DB.RecentEvents(
-		ctx, roomID, r, numRecentEventsPerRoom, true, true,
+		ctx, roomID, r, filter.Room.Timeline.Limit, true, true,
 	)
 	if err != nil {
 		return
 	}
-
-	// TODO FIXME: We don't fully implement history visibility yet. To avoid leaking events which the
-	// user shouldn't see, we check the recent events and remove any prior to the join event of the user
-	// which is equiv to history_visibility: joined
-	joinEventIndex := -1
-	for i := len(recentStreamEvents) - 1; i >= 0; i-- {
-		ev := recentStreamEvents[i]
-		if ev.Type() == gomatrixserverlib.MRoomMember && ev.StateKeyEquals(device.UserID) {
-			membership, _ := ev.Membership()
-			if membership == "join" {
-				joinEventIndex = i
-				if i > 0 {
-					// the create event happens before the first join, so we should cut it at that point instead
-					if recentStreamEvents[i-1].Type() == gomatrixserverlib.MRoomCreate && recentStreamEvents[i-1].StateKeyEquals("") {
-						joinEventIndex = i - 1
-						break
-					}
-				}
-				break
-			}
+	recentStreamEvents, limited = p.filterStreamEventsAccordingToHistoryVisibility(recentStreamEvents, stateEvents, device, limited)
+
+	// TODO: How do we apply filter.Room.State to this as well?
+	// Is there a generic function where we can pass a list of events and filter and get the result?
+	// This is based off of Synapse where we derive the state from the resultant timeline events
+	// https://github.com/matrix-org/synapse/blob/14950a45d6ff3a5ea737322af1096a49b079f2eb/synapse/handlers/sync.py#L791-L795
```

Comment:
@neilalexander Do you know of a utility function to accomplish this?

Reply:
Filtering when we select the events is probably the way to go. Related to https://github.com/matrix-org/dendrite/pull/1671#discussion_r559699059